txnsync/peer.go

```diff
@@ -0,0 +1,809 @@
+	// "snapshot" from the transaction pool, and send that "snapshot" to completion before attempting to re-iterate.
+	messageSeriesPendingTransactions []transactions.SignedTxGroup
+
+	// transactionPoolAckCh is passed to the transaction handler when incoming transaction arrives. The channel is passed upstream, so that once
+	// a transaction is added to the transaction pool, we can get some feedback for that.
+	transactionPoolAckCh chan uint64
+
+	// transactionPoolAckMessages maintain a list of the recent incoming messages sequence numbers whose transactions were added fully to the transaction
+	// pool. This list is being flushed out every time we send a message to the peer.
+	transactionPoolAckMessages []uint64
+
+	// used by the selectPendingTransactions method, the lastSelectedTransactionsCount contains the number of entries selected on the previous iteration.
+	// this value is used to optimize the memory preallocation for the selection IDs array.
+	lastSelectedTransactionsCount int
+}
+
+// requestParamsGroupCounterState stores the latest group counters for a given set of request params.
+// we use this to ensure we can have multiple iteration of bloom filter scanning over each individual
+// transaction group. This method allow us to reduce the bloom filter errors while avoid scanning the
+// list of transactions redundently.
+//msgp:ignore transactionGroupCounterState
+type requestParamsGroupCounterState struct {
+	offset        byte
+	modulator     byte
+	groupCounters [bloomFilterRetryCount]uint64
+}
+
+// transactionGroupCounterTracker manages the group counter state for each request param.
+//msgp:ignore transactionGroupCounterTracker
+type transactionGroupCounterTracker []requestParamsGroupCounterState
+
+// get returns the group counter for a given set of request param.
+func (t *transactionGroupCounterTracker) get(offset, modulator byte) uint64 {
+	i := t.index(offset, modulator)
+	if i >= 0 {
+		return (*t)[i].groupCounters[0]
+	}
+	return 0
+}
+
+// set updates the group counter for a given set of request param. If no such request
+// param currently exists, it create it.
+func (t *transactionGroupCounterTracker) set(offset, modulator byte, counter uint64) {
+	i := t.index(offset, modulator)
+	if i >= 0 {
+		(*t)[i].groupCounters[0] = counter
+		return
+	}
+	// if it doesn't exists -
+	state := requestParamsGroupCounterState{
+		offset:    offset,
+		modulator: modulator,
+	}
+	state.groupCounters[0] = counter
+
+	if len(*t) == maxTransactionGroupTrackers {
+		// shift all entries by one.
+		copy((*t)[0:], (*t)[1:])
+		(*t)[maxTransactionGroupTrackers-1] = state
+	} else {
+		*t = append(*t, state)
+	}
+}
+
+// roll the counters for a given requests params, so that we would go back and
+// rescan some of the previous transaction groups ( but not all !) when selectPendingTransactions is called.
+func (t *transactionGroupCounterTracker) roll(offset, modulator byte) {
```

Comment:
I am not following how roll is working.
Is there a limit on how many times it can be called to preserve the remembered information?

Reply:
the roll feature is pretty simple : it allows us to quantify how many bloom filters ( that belongs to the same offset/modulator ) would be tested against each transaction group in the worst case scenario. It was implemented to combat false-positives on a low-precision bloom filters, while allowing us to limit the number of retries.

With the current 32bit xor bloom filters, it's needed only since we have a fallback to 8 bit multiset hash.